gio/src/socket.rs

```diff
@@ -61,6 +69,185 @@ impl AsRawSocket for Socket {
+    len: u32,
+}
+
+impl SocketControlMessages {
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            ptr: ptr::null_mut(),
+            len: 0,
+        }
+    }
+}
+
+impl AsRef<[SocketControlMessage]> for SocketControlMessages {
+    #[inline]
+    fn as_ref(&self) -> &[SocketControlMessage] {
+        unsafe { std::slice::from_raw_parts(self.ptr as *const _, self.len as usize) }
+    }
+}
+
+impl std::ops::Deref for SocketControlMessages {
+    type Target = [SocketControlMessage];
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl Default for SocketControlMessages {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Drop for SocketControlMessages {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            let _: Slice<SocketControlMessage> =
+                Slice::from_glib_full_num(self.ptr as *mut _, self.len as usize);
+        }
+    }
+}
+
+#[doc(alias = "GInputMessage")]
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct InputMessage<'m> {
+    message: ffi::GInputMessage,
+    address: PhantomData<Option<&'m mut Option<SocketAddress>>>,
+    vectors: PhantomData<&'m mut [InputVector<'m>]>,
+    control_messages: PhantomData<Option<&'m mut SocketControlMessages>>,
+}
+
+impl<'m> InputMessage<'m> {
+    pub fn new(
+        mut address: Option<&'m mut Option<SocketAddress>>,
+        vectors: &'m mut [InputVector<'m>],
+        control_messages: Option<&'m mut SocketControlMessages>,
+    ) -> Self {
+        let address = address
+            .as_mut()
+            .map(|a| {
+                assert!(a.is_none());
+                *a as *mut _ as *mut _
+            })
+            .unwrap_or_else(ptr::null_mut);
+        let (control_messages, num_control_messages) = control_messages
+            .map(|c| (&mut c.ptr as *mut _, &mut c.len as *mut _))
+            .unwrap_or_else(|| (ptr::null_mut(), ptr::null_mut()));
+        Self {
+            message: ffi::GInputMessage {
+                address,
+                vectors: vectors.as_mut_ptr() as *mut ffi::GInputVector,
+                num_vectors: vectors.len().try_into().unwrap(),
+                bytes_received: 0,
+                flags: 0,
+                control_messages,
+                num_control_messages,
+            },
+            address: PhantomData,
+            vectors: PhantomData,
+            control_messages: PhantomData,
+        }
+    }
+    #[inline]
+    pub const fn flags(&self) -> i32 {
+        self.message.flags
+    }
+    #[inline]
+    pub const fn bytes_received(&self) -> usize {
+        self.message.bytes_received
+    }
+}
+
+#[doc(alias = "GOutputVector")]
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct OutputVector<'v> {
+    _vector: ffi::GOutputVector,
+    buffer: PhantomData<&'v [u8]>,
+}
+
+impl<'v> OutputVector<'v> {
+    #[inline]
+    pub const fn new(buffer: &'v [u8]) -> Self {
```

Comment:
For completeness etc we could `impl Deref<Target=[u8]>` here and/or provide getters for the two fields. Also for the `InputVector` and input/output messages

Reply:
Added `Deref` impls and getters for `vectors`